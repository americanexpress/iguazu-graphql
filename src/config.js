/*
 * Copyright 2018 American Express
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const endpoints = new Map();

// function addEndpoint({ name, ...opts }) {
function addEndpoint(name, opts) {
  // const { name } = opts;
  if (!name || typeof name !== 'string') {
    throw new Error(`name must be a string (was ${name ? typeof name : 'falsey'})`);
  }

  if (endpoints.has(name)) {
    throw new Error(`cannot override existing config for ${name}`);
  }

  endpoints.set(name, opts);
}

function getEndpoint(name) {
  if (!endpoints.has(name)) {
    throw new Error(`${name} is not a configured endpoint`);
  }
  // return {
  //   ...endpoints.get(name),
  //   name,
  // };
  return endpoints.get(name);
}

const config = {
  baseFetch: fetch,
  // don't like the name, but it's part of iguazu-rest so it's consistent ðŸ˜¬
  getToState: state => state.iguazuGraphQL,
  addEndpoint,
  getEndpoint,
};

export function configureIguazuGraphQL(customConfig) {
  Object.assign(config, customConfig, { addEndpoint, getEndpoint });
}

export default config;
