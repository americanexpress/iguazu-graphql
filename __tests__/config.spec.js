/*
 * Copyright 2018 American Express
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import 'fetch-everywhere';

import config, { configureIguazuGraphQL } from '../src/config';

describe('config', () => {
  const originalGetToState = config.getToState;
  const originalBaseFetch = config.baseFetch;

  function resetConfig() {
    config.getToState = originalGetToState;
    config.baseFetch = originalBaseFetch;
  }

  it('uses the global fetch as the default baseFetch', () => {
    expect(config).toHaveProperty('baseFetch', global.fetch);
  });

  it('looks for iguazuGraphQL as the default getToState redux state branch', () => {
    expect(config).toHaveProperty('getToState', expect.any(Function));
    const stateBranch = {};
    const state = { iguazuGraphQL: stateBranch };
    expect(config.getToState(state)).toBe(stateBranch);
  });

  describe('addEndpoint', () => {
    it('throws when not provided with a name to reference the endpoint configuration', () => {
      expect(() => config.addEndpoint(null)).toThrowErrorMatchingSnapshot();
      expect(() => config.addEndpoint(6)).toThrowErrorMatchingSnapshot();
    });

    it('throws when provided with an existing name', () => {
      expect(() => config.addEndpoint('throws-when-provided-with-an-existing-name')).not.toThrowError();
      expect(() => config.addEndpoint('throws-when-provided-with-an-existing-name')).toThrowErrorMatchingSnapshot();
    });
  });

  describe('getEndpoint', () => {
    it('throws when provided with an name that does not exist', () => {
      expect(() => config.getEndpoint('throws-when-provided-with-an-name-that-does-not-exist')).toThrowErrorMatchingSnapshot();
    });
  });

  test('addEndpoint makes an endpoint configuration available to getEndpoint by the name', () => {
    const opts = { anOption: 'a value' };
    config.addEndpoint('addEndpoint-makes-an-endpoint-configuration-available-to-getEndpoint-by-the-name', opts);
    expect(config.getEndpoint('addEndpoint-makes-an-endpoint-configuration-available-to-getEndpoint-by-the-name')).toBe(opts);
  });

  describe('configureIguazuGraphQL', () => {
    beforeEach(resetConfig);

    it('changes baseFetch', () => {
      const baseFetch = jest.fn();
      configureIguazuGraphQL({ baseFetch });
      expect(config.baseFetch).toBe(baseFetch);
    });

    it('changes getToState', () => {
      const getToState = jest.fn();
      configureIguazuGraphQL({ getToState });
      expect(config.getToState).toBe(getToState);
    });

    it('cannot change addEndpoint', () => {
      const addEndpoint = jest.fn();
      configureIguazuGraphQL({ addEndpoint });
      expect(config.addEndpoint).not.toBe(addEndpoint);
    });

    it('cannot change getEndpoint', () => {
      const getEndpoint = jest.fn();
      configureIguazuGraphQL({ getEndpoint });
      expect(config.getEndpoint).not.toBe(getEndpoint);
    });
  });
});
