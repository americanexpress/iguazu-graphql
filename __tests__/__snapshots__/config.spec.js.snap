// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config addEndpoint throws when not provided with a name to reference the endpoint configuration 1`] = `"name must be a string (was falsey)"`;

exports[`config addEndpoint throws when not provided with a name to reference the endpoint configuration 2`] = `"name must be a string (was number)"`;

exports[`config addEndpoint throws when provided with an existing name 1`] = `"cannot override existing config for throws-when-provided-with-an-existing-name"`;

exports[`config getEndpoint throws when provided with an name that does not exist 1`] = `"throws-when-provided-with-an-name-that-does-not-exist is not a configured endpoint"`;
